File: crates/perplex-matching/src/order.rs
use perplex_core::types::{Address, MarketId, OrderId, TsNanos};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};

use crate::book::Side;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum OrderType {
    Market,
    Limit,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum TimeInForce {
    /// Good-till-cancelled (default for limits).
    Gtc,
    /// Immediate-or-cancel: cross what's available now, cancel any unfilled remainder.
    Ioc,
    /// Fill-or-kill: cross the FULL qty atomically or reject entirely.
    Fok,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Order {
    pub id: OrderId,
    pub owner: Address,
    pub market: MarketId,
    pub side: Side,
    pub order_type: OrderType,
    /// For market orders this is ignored; we still keep it so a market order can carry
    /// a slippage cap if the caller wants one (encoded as a finite price).
    pub price: Decimal,
    pub qty: Decimal,
    pub remaining: Decimal,
    pub time_in_force: TimeInForce,
    pub post_only: bool,
    pub reduce_only: bool,
    pub ts_ns: TsNanos,
}

impl Order {
    /// True if `self` (a taker) would cross `maker_price`.
    pub fn crosses(&self, maker_price: Decimal) -> bool {
        match (self.side, self.order_type) {
            (Side::Buy, OrderType::Market) => true,
            (Side::Sell, OrderType::Market) => true,
            (Side::Buy, OrderType::Limit) => self.price >= maker_price,
            (Side::Sell, OrderType::Limit) => self.price <= maker_price,
        }
    }
}

File: web/lib/eip712/order.ts
import { keccak256, toHex, parseUnits } from "viem";
import type { MarketId, Side, OrderType, TimeInForce } from "@/lib/types/contract";

export const ORDER_DOMAIN = (chainId: number, verifyingContract: `0x${string}`) =>
  ({
    name: "Perplex",
    version: "1",
    chainId,
    verifyingContract,
  }) as const;

export const ORDER_TYPES = {
  Order: [
    { name: "owner", type: "address" },
    { name: "marketId", type: "bytes32" },
    { name: "side", type: "uint8" },
    { name: "orderType", type: "uint8" },
    { name: "price", type: "uint256" },
    { name: "qty", type: "uint256" },
    { name: "timeInForce", type: "uint8" },
    { name: "reduceOnly", type: "bool" },
    { name: "postOnly", type: "bool" },
    { name: "nonce", type: "uint256" },
    { name: "expiryTsSec", type: "uint64" },
  ],
} as const;

const SIDE: Record<Side, number> = { buy: 0, sell: 1 };
const TYPE: Record<OrderType, number> = { market: 0, limit: 1 };
const TIF: Record<TimeInForce, number> = { gtc: 0, ioc: 1, fok: 2 };

export function marketIdHash(marketId: MarketId): `0x${string}` {
  return keccak256(toHex(marketId));
}

export interface BuildOrderInput {
  owner: `0x${string}`;
  marketId: MarketId;
  side: Side;
  type: OrderType;
  price: string;
  qty: string;
  timeInForce: TimeInForce;
  reduceOnly: boolean;
  postOnly: boolean;
  nonce: string;
  expiryTsSec?: number;
}

export function buildOrderTypedMessage(input: BuildOrderInput) {
  return {
    owner: input.owner,
    marketId: marketIdHash(input.marketId),
    side: SIDE[input.side],
    orderType: TYPE[input.type],
    price: parseUnits(input.price, 18),
    qty: parseUnits(input.qty, 18),
    timeInForce: TIF[input.timeInForce],
    reduceOnly: input.reduceOnly,
    postOnly: input.postOnly,
    nonce: BigInt(input.nonce),
    expiryTsSec: BigInt(input.expiryTsSec ?? 0),
  };
}
